### PR TITLE
[MM-52268] Add initial support for connecting through TCP candidates

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

PR adds initial support for generating and allowing connection through TCP candidates. This allows the service to be self sufficient even in cases where clients cannot establish UDP connections and can avoid the complexity of deploying dedicated TURN servers.

[Initial load-tests](https://github.com/mattermost/rtcd/assets/1832946/672bc3a6-4cb7-405e-a279-b67ed9191dc8) didn't show a significant difference in performance. So far I tested TCP only vs UDP only. Would be good to test a more realistic scenario with both candidate types used at the same time.

Some current blockers:

- More performance tests need to be conducted.
- ~~There's [an issue](https://github.com/pion/ice/issues/583) preventing connectivity when listening on all available local interfaces. Still need to understand if it's a configuration issue on our part or a bug in `pion/ice` but so far I couldn't figure it out.~~
  - This was resolved by simplifying the logic in [b9665a2](https://github.com/mattermost/rtcd/pull/103/commits/b9665a23c926b94a750f24ac3294980843a83827)
- ~~There's an issue when shutting down the server, listeners take a long time to be closed causing sporadic timeouts. Needs further investigation.~~
  - This is the API server graceful shutdown which can take a bit when having long lasting connections (e.g. fetching pprof profiles).
- Documentation updates.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-52268

### CODE REVIEW

File: service/rtc/utils.go [hunk 1]
```diff
@@ -33,38 +34,63 @@ func generateAddrsPairs(localIPs []string, publicAddrsMap map[string]string, hos
 	var pairs []string
 	var hostOverrideIP string
 
+	// If the override is in full NAT mapping format (e.g. "EA/IA,EB/IB") we return
```
Comment (cpoile) [inline service/rtc/utils.go:37]: I can't seem to find what the nat mapping format is, can you give me a link to read up on it? I'm guessing it means: external addr/internal addr,external binding/internal binding?

File: service/rtc/utils.go [hunk 2]
```diff
@@ -33,38 +34,63 @@ func generateAddrsPairs(localIPs []string, publicAddrsMap map[string]string, hos
...
-		publicAddr := publicAddrsMap[localAddr]
+	// Nothing to do at this point if no local IP was found.
+	if len(localIPs) == 0 {
+		return nil, nil
+	}
 
-		// If an override was explicitly provided we enforce that.
-		if hostOverrideIP != "" {
-			publicAddr = hostOverrideIP
+	// If the override is set but no explicit mapping is given, we try to
```
Comment (cpoile) [inline service/rtc/utils.go:56]: these are helpful comments, thank you

**Review by cpoile — DISMISSED** [review]
Very cool, love it. Questions for my own education.

File: config/config.sample.toml [hunk 3]
```diff
@@ -23,15 +23,22 @@ security.admin_secret_key = ""
...
 [rtc]
-# The IP address used to listen for UDP packets. If left empty it has the same
+# The IP address used to listen for UDP packets and generate UDP candidates. If left empty it has the same
 # effect as using the 0.0.0.0 address, causing the server to listen on all available network interfaces.
 ice_address_udp = ""
-# The UDP port used to route all the media (audio/screen/video tracks).
+# The UDP port used to route media (audio/screen/video tracks).
 ice_port_udp = 8443
+# The IP address used to listen for TCP connections and generate TCP candidates. This is used to generate
+# TCP candidates which may be used by client in case UDP connectivity is not available.
```
[the remark this answers is not in the document]
Comment (streamer45) [reply config/config.sample.toml:32]: Yes, that's the behaviour :+1:

File: service/rtc/utils.go (hunk 1 above)
Comment (streamer45) [reply service/rtc/utils.go:37]: Oh, letters are purely sequentials, you could have EC/IC,ED/ID and so on. These are pairs of external IP / internal IP mappings.

It's not really well documented in https://pkg.go.dev/github.com/pion/webrtc/v3#SettingEngine.SetNAT1To1IPs I had to ask to find out how to properly use it.

Comment (streamer45) [reply service/rtc/utils.go:37]: I updated the config documentation slightly. Hopefully it's clearer now.

**Review by cpoile — DISMISSED** [review]

**Review by cpoile — DISMISSED** [review]

**Comment by streamer45** [comment]
@cpoile Apologies but had to push some small docs changes.

**Review by cpoile — APPROVED** [review]

**Comment by cpoile** [comment]
Maybe we can ask to have this repo exempted from the stale review rule?

**Pull request merged**